flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/dynamicfiltering/DynamicFilteringDataCollectorOperator.java

```diff
@@ -0,0 +1,183 @@
+        this.operatorEventGateway = checkNotNull(operatorEventGateway);
+    }
+
+    @Override
+    public void open() throws Exception {
+        super.open();
+        this.typeInfo = InternalTypeInfo.of(dynamicFilteringFieldType);
+        this.serializer = typeInfo.createSerializer(new ExecutionConfig());
+        this.buffer = new TreeSet<>(new BytePrimitiveArrayComparator(true)::compare);
+        this.currentSize = 0L;
+    }
+
+    @Override
+    public void processElement(StreamRecord<RowData> element) throws Exception {
+        if (exceedThreshold()) {
+            return;
+        }
+
+        RowData value = element.getValue();
+        GenericRowData rowData = new GenericRowData(dynamicFilteringFieldIndices.size());
+        for (int i = 0; i < dynamicFilteringFieldIndices.size(); ++i) {
+            LogicalType type = dynamicFilteringFieldType.getTypeAt(i);
+            int index = dynamicFilteringFieldIndices.get(i);
+            switch (type.getTypeRoot()) {
+                case INTEGER:
+                    rowData.setField(i, value.getInt(index));
+                    break;
+                case BIGINT:
+                    rowData.setField(i, value.getLong(index));
+                    break;
+                case VARCHAR:
+                    rowData.setField(i, value.getString(index));
+                    break;
+                default:
+                    throw new UnsupportedOperationException(
+                            "Unsupported data type for dynamic filtering:" + type);
+            }
+        }
+
+        try (ByteArrayOutputStream baos = new ByteArrayOutputStream()) {
+            DataOutputViewStreamWrapper wrapper = new DataOutputViewStreamWrapper(baos);
+            serializer.serialize(rowData, wrapper);
+            boolean duplicated = !buffer.add(baos.toByteArray());
+            if (duplicated) {
+                return;
+            }
+
+            currentSize += baos.size();
+        }
+
+        if (exceedThreshold()) {
+            // Clear the filtering data and disable self by leaving the currentSize unchanged
+            buffer.clear();
+            LOG.warn(
+                    "Collected data size exceeds the threshold, {} > {}, dynamic filtering is disabled.",
+                    currentSize,
+                    threshold);
+        }
+    }
+
+    private boolean exceedThreshold() {
+        return threshold > 0 && currentSize > threshold;
+    }
+
+    @Override
+    public void finish() throws Exception {
+        if (exceedThreshold()) {
+            LOG.info(
+                    "Finish collecting. {} bytes are collected which exceeds the threshold {}. Sending empty data.",
+                    currentSize,
+                    threshold);
+        } else {
+            LOG.info(
+                    "Finish collecting. {} bytes in {} rows are collected. Sending the data.",
+                    currentSize,
+                    buffer.size());
+        }
+        sendEvent();
```

Comment:
what if we do not send any event when `exceedThreshold` ?

Reply:
In case different attempts of the collector calculate different results, one of which is exceeding threshold, I think it's better to send the event even it's exceeding threshold, or the coordinator may not know the problem and can't trigger failover for recomputing.